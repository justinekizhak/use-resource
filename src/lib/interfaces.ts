import {
  MutableRefObject,
  ReactChild,
  ReactFragment,
  ReactPortal
} from "react";
import {
  AxiosError,
  AxiosInstance,
  AxiosRequestConfig,
  AxiosResponse
} from "axios";

export interface DebugObject {
  timestamp: string;
  message?: string;
  data?: object;
}

export type JsxComponentType =
  | boolean
  | ReactChild
  | ReactFragment
  | ReactPortal
  | JSX.Element
  | null
  | undefined;

export type ResourceType = {
  data: object;
  isLoading: boolean;
  errorData: object | null | undefined;
  debug: MutableRefObject<DebugObject[]>;
  cancel: any;
  refetch: (customConfig: BaseConfigType) => void;
};

export type ResourceContextState = {
  [key: string]: ResourceType;
};

export type ResourceContextType = ResourceContextState | null | undefined;

export type UseResourceBasicOptionsType = {
  CustomContext?: React.Context<ResourceContextType> | null | undefined;
  triggerOn?: string | boolean | any[];
  onMountCallback?: (customAxios: AxiosInstance) => void;
};

export type LoadingComponentType = () => JSX.Element;
export type ErrorComponentType = (
  errorMessage: string,
  errorData: any
) => JSX.Element;

export type UseResourceAdvancedOptionsType = {
  globalLoadingComponent?: LoadingComponentType;
  globalErrorComponent?: ErrorComponentType;
  useMessageQueue?: boolean | object;
  useGlobalContext?: boolean;
  devMode?: boolean;
};

export type GlobalResourceContextType = {
  dispatch: (data: ResourceContextType) => void;
  selector: (data: any) => any;
};

export interface ContextContainerPropsType {
  children?: JsxComponentType;
  loadingComponent?: LoadingComponentType;
  errorComponent?: ErrorComponentType;
}

export type ContextContainerType = (
  props: ContextContainerPropsType
) => JSX.Element;

export interface UseResourceReturnType extends ResourceType {
  Container: ContextContainerType;
}

export type BaseConfigType = AxiosRequestConfig | ChainedRequestConfigType[];

export type UseResourceType = (
  defaultConfig: BaseConfigType,
  resourceName: string,
  options?: UseResourceOptionsType
) => UseResourceReturnType;

export type ChainResponseType = object | AxiosResponse | void;
export type AccumulatorType = (object | AxiosResponse)[];
export type AccumulatorContainer = { current: AccumulatorType };
export type NextType = (data: ChainResponseType) => AccumulatorContainer;

export type BeforeTaskType = (
  accumulator?: AccumulatorContainer,
  next?: NextType,
  disableStateUpdate?: boolean
) => AxiosRequestConfig | void;

export type TaskType = (
  customConfig: AxiosRequestConfig,
  accumulator?: AccumulatorContainer,
  next?: NextType
) => Promise<ChainResponseType>;

export type OnSuccessType = (
  response: ChainResponseType,
  accumulator?: AccumulatorContainer,
  next?: NextType,
  disableStateUpdate?: boolean
) => ChainResponseType;

export type OnFailureType = (
  error: any | AxiosError,
  accumulator?: AccumulatorContainer,
  next?: NextType
) => ChainResponseType;

export type OnFinalType = (
  accumulator?: AccumulatorContainer,
  next?: NextType,
  disableStateUpdate?: boolean
) => ChainResponseType;

export interface ChainedRequestConfigType extends Object {
  baseConfig: AxiosRequestConfig;
  beforeTask?: BeforeTaskType;
  task?: TaskType;
  onSuccess?: OnSuccessType;
  onFailure?: OnFailureType;
  onFinal?: OnFinalType;
}

export interface UseResourceOptionsType
  extends UseResourceBasicOptionsType,
    UseResourceAdvancedOptionsType {}

export type DispatchCallbackType = (state: ResourceContextState) => any;

export type PushToAccumulatorType = (
  next: NextType | undefined,
  res: ChainResponseType | undefined
) => void;
