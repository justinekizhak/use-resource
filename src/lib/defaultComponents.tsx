import React from "react";

import { ErrorComponentType, LoadingComponentType } from "./interfaces";

export const defaultLoadingComponent: LoadingComponentType = () => (
  <div className="loading"> Loading... </div>
);
export const defaultErrorComponent: ErrorComponentType = (
  errorMessage: string,
  errorData: any
) => <div className="error-message"> {errorMessage} </div>;
