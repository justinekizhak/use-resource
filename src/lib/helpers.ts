import { MutableRefObject } from "react";
import { AxiosRequestConfig, AxiosResponse, AxiosError } from "axios";

import {
  OnSuccessType,
  BeforeTaskType,
  TaskType,
  OnFailureType,
  OnFinalType,
  ChainedRequestConfigType,
  BaseConfigType,
  PushToAccumulatorType
} from "./interfaces";

export const getTriggerDependencies = (
  triggerOn: string | boolean | any[] = "onMount",
  axiosConfig: AxiosRequestConfig = {}
): [any[], boolean] => {
  // Highest priority is if the triggerOn is an array
  if (Array.isArray(triggerOn)) {
    return [triggerOn, true];
  }
  // Second priority is if the triggerOn is a boolean
  if (triggerOn === false) {
    return [[], false];
  }
  if (triggerOn === true) {
    return [[], true];
  }
  // Third priority is if the triggerOn is a string
  if (triggerOn === "onMount") {
    return [[], true];
  }
  // Fourth priority is when request is a GET request
  if (axiosConfig.method === "get") {
    return [[], false];
  }
  // By default, on mount trigger is false
  return [[], false];
};

export const getMessageQueueData = (data: boolean | object = false) => {
  return [false, ""];
  // if (typeof data === "object") {
  //   const isAvailable = true;
  //   const keyName = data?.keyName || "";
  //   return [isAvailable, keyName];
  // }
  // const isAvailable = data;
  // const keyName = `${Date.now()}`;
  // return [isAvailable, keyName];
};

export const getBaseConfig = (
  baseConfig: BaseConfigType,
  index = 0
): AxiosRequestConfig => {
  const useRequestChaining = Array.isArray(baseConfig);
  if (useRequestChaining && baseConfig.length === 0) {
    throw new Error("Please pass in the request config");
  }
  const defaultConfig: AxiosRequestConfig = useRequestChaining
    ? baseConfig[index]["baseConfig"]
    : baseConfig;

  return defaultConfig;
};

export const getErrorMessage = (
  errorData: AxiosError | AxiosResponse | undefined
): string => {
  const defaultErrorMessage = "Something went wrong. Please try again.";
  const err = errorData as AxiosError;
  if (err?.response?.data?.message) {
    return err.response.data.message || defaultErrorMessage;
  }
  if (err?.message) {
    return err.message || defaultErrorMessage;
  }
  return "";
};

export const getFunc = (
  requestObject: ChainedRequestConfigType,
  key: string
) => {
  const func =
    // @ts-ignore
    requestObject && typeof requestObject[key] === "function"
      ? // @ts-ignore
        requestObject[key]
      : () => {};
  return func;
};

export const pushToAcc: PushToAccumulatorType = (next, res) => {
  if (next && typeof next === "function") {
    if (res) {
      next(res);
    }
  }
};

export const getFinalRequestChain = (
  newChainedRequestData: ChainedRequestConfigType,
  index: number,
  fullBaseConfigList: ChainedRequestConfigType[],
  beforeTask: BeforeTaskType,
  task: TaskType,
  onSuccess: OnSuccessType,
  onFailure: OnFailureType,
  onFinal: OnFinalType,
  controllerInstance: MutableRefObject<AbortController> | undefined = undefined
): ChainedRequestConfigType => {
  const oldChainedRequestData = fullBaseConfigList[index];
  const finalConfig = {
    ...oldChainedRequestData["baseConfig"],
    ...newChainedRequestData["baseConfig"]
  };
  // The new beforeTask will overwrite the old beforeTask
  const finalBeforeTask: BeforeTaskType = (acc, next) => {
    const func: BeforeTaskType = getFunc(newChainedRequestData, "beforeTask");
    const res = func(acc, next);
    pushToAcc(next, res);
    const res2 = beforeTask(acc, next);
    pushToAcc(next, res2);
    return res2;
  };

  // The new task will overwrite all the task
  const finalTask: TaskType = async (customConfig, acc, next) => {
    const func: TaskType = getFunc(newChainedRequestData, "task");
    const config1 = {
      signal: controllerInstance?.current?.signal,
      ...finalConfig,
      ...customConfig
    };
    const res: AxiosRequestConfig = (await func(config1, acc, next)) || config1;
    pushToAcc(next, res);
    const res2 = await task(res, acc, next);
    pushToAcc(next, res2);
    return res2;
  };

  // Runs the request through the user callback then the response is sent to the actual success task
  const finalOnSuccess: OnSuccessType = (
    res,
    acc,
    next,
    disableStateUpdate
  ) => {
    const func: OnSuccessType = getFunc(newChainedRequestData, "onSuccess");
    const res2 = func(res, acc, next, disableStateUpdate) || res;
    pushToAcc(next, res2);
    const res3 = onSuccess(res2, acc, next, disableStateUpdate);
    pushToAcc(next, res3);
    return res3;
  };

  const finalOnFailure: OnFailureType = (error, acc, next) => {
    const func: OnFailureType = getFunc(newChainedRequestData, "onFailure");
    const res = func(error, acc, next) || error;
    pushToAcc(next, res);
    const res2 = onFailure(res, acc, next);
    pushToAcc(next, res2);
    return res2;
  };

  const finalOnFinal: OnFinalType = (acc, next) => {
    const func: OnFinalType = getFunc(newChainedRequestData, "onFinal");
    const res = func(acc, next);
    pushToAcc(next, res);
    const res2 = onFinal(acc, next);
    pushToAcc(next, res2);
    return res2;
  };

  return {
    baseConfig: finalConfig,
    beforeTask: finalBeforeTask,
    task: finalTask,
    onSuccess: finalOnSuccess,
    onFailure: finalOnFailure,
    onFinal: finalOnFinal
  };
};
