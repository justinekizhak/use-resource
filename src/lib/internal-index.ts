/**
 * Here contains only functions and hooks that are used both internally and exported in the library.
 *
 * Excludes all the types infos.
 *
 * For info on the exported functions, see {@link Externals} module.
 *
 * For info on the types, see {@link Types} module.
 *
 * @module Internals
 */

// requestChain
export * from "./requestChain/dependencyMap";
export * from "./requestChain/handler";
export * from "./requestChain/parallelApiHandler";

// resourceContext
export * from "./resourceContext/context";
export * from "./resourceContext/handleEvents";
export * from "./resourceContext/hooks";
export * from "./resourceContext/provider";

// utils
export * from "./utils/defaultComponents";
export * from "./utils/helpers";
export * from "./utils/refetch";
export * from "./utils/requestChain";

export * from "./useResource";
