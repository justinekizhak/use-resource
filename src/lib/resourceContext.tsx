import { createContext } from "react";
import { GlobalResourceContextType } from "./interfaces";

export const GlobalResourceContext = createContext<GlobalResourceContextType>({
  dispatch: () => {
    console.info("Please add a global resource context provider.");
  },
  selector: () => {
    console.info("Please add a global resource context provider.");
  }
});
