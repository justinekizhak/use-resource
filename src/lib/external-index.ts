/**
 * Only the functions and hooks exported here are available to be used public.
 *
 * Rest all are under the {@link Internals} module.
 *
 * For type information, see {@link Types} module.
 *
 * @module Externals
 */

import {
  GlobalResourceContext,
  generateContextDefault
} from "./resourceContext/context";
import { GlobalResourceContextProvider } from "./resourceContext/provider";

export { useResource } from "./useResource";

/**
 * GlobalResource object.
 *
 * Use this object when you want to add the global-resource-provider or use its corresponding context.
 *
 * You can also create your own global-resource context and use it.
 *
 * For creating your own global-resource context, see {@link generateContextDefault}.
 */
export const GlobalResource = {
  Context: GlobalResourceContext,
  Provider: GlobalResourceContextProvider
};

export { generateContextDefault };

export { useSelector, useDispatch, usePublish } from "./resourceContext/hooks";
