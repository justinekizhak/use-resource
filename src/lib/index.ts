import { GlobalResourceContext } from "./resourceContext/context";
import { GlobalResourceContextProvider } from "./resourceContext/provider";

export { useResource } from "./useResource";

export const GlobalResource = {
  Context: GlobalResourceContext,
  Provider: GlobalResourceContextProvider
};

export {
  useSelector,
  useDispatch,
  usePublish
} from "lib/resourceContext/hooks";
