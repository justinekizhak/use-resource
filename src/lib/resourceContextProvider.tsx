import React, { useState, useCallback } from "react";

import {
  JsxComponentType,
  ResourceContextType,
  ResourceContextState,
  DispatchCallbackType
} from "./interfaces";

import { GlobalResourceContext } from "./resourceContext";

export const GlobalResourceContextProvider = (props: {
  children: JsxComponentType;
}) => {
  const [state, setState] = useState<ResourceContextState>({});

  const dispatch = useCallback((inputData: ResourceContextType) => {
    if (!inputData) {
      return;
    }
    setState((oldData) => {
      const newData = { ...oldData, ...inputData };
      return newData;
    });
  }, []);

  const selector = useCallback(
    (callback): DispatchCallbackType => {
      const res = callback(state);
      return res;
    },
    [state]
  );

  return (
    <GlobalResourceContext.Provider value={{ dispatch, selector }}>
      {props.children}
    </GlobalResourceContext.Provider>
  );
};
