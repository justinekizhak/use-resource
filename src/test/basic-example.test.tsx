import { useResource } from "../lib";
import { useRenderCount } from "../examples/utils/useRenderCount";
import {
  render,
  screen,
  getNodeText,
  waitForElementToBeRemoved
} from "@testing-library/react";
import snapshotDiff from "snapshot-diff";

const Wrapper = () => {
  const { RenderContainer } = useRenderCount();
  const { data, Container } = useResource({
    url: "https://test.com/test/1"
  });

  return (
    <>
      <RenderContainer />
      <Container>
        <div>{JSON.stringify(data)}</div>
      </Container>
    </>
  );
};

const setup = () => {
  return render(<Wrapper />);
};

describe("basic-example", () => {
  it("Check render count", async () => {
    setup();
    const counterEl = await screen.findByTitle("render-counter");
    const value = getNodeText(counterEl);
    expect(value).toBe("2");
  });
  it("Check data content", async () => {
    const { asFragment } = setup();
    const initialRender = asFragment();
    await waitForElementToBeRemoved(screen.getByText("Loading..."));
    const afterLoading = asFragment();
    expect(snapshotDiff(initialRender, afterLoading)).toMatchSnapshot();
  });
});
