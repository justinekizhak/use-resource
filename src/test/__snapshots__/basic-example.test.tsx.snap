// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic-example Check data content 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -7,24 +7,21 @@
      >
        Total render count(including initial render):
        <span
          title=\\"render-counter\\"
        >
-         2
+         4
        </span>
      </div>
    </div>
    <div
      class=\\"resource-resource\\"
    >
      <div
        class=\\"content\\"
      >
-       <div
-         class=\\"loading\\"
-       >
-          Loading... 
+       <div>
+         {\\"hello\\":\\"world\\"}
        </div>
-       <div />
      </div>
    </div>
  </DocumentFragment>"
`;
